src/SSHDebugPS/RawShell.cs

```diff
@@ -0,0 +1,295 @@
+            _stdoutWriter.Flush();
+        }
+
+        public void WriteCommandStart(string startCommand)
+        {
+            WriteLine(startCommand);
+        }
+
+        public void WriteLine(string text)
+        {
+            _stdoutWriter.WriteLine(text);
+            _stdoutWriter.Flush();
+        }
+
+        private void OnErrorOutput(object sender, DataReceivedEventArgs e)
+        {
+            ErrorOccured?.Invoke(sender, e);
+        }
+
+        private void OnProcessOutput(object sender, DataReceivedEventArgs e)
+        {
+            OutputReceived?.Invoke(this, e?.Data + '\n');
+        }
+
+        private void OnProcessExited(object sender, EventArgs e)
+        {
+            Closed?.Invoke(this, _localProcess.ExitCode);
+        }
+    }
+
+    /// <summary>
+    /// Launches a raw command that doesn't have shell-like input/output
+    /// </summary>
+    internal class CommandShell : IRawShell
+    {
+        private System.Diagnostics.Process _localProcess;
+        private StreamWriter _stdoutWriter;
+        private StreamReader _processReader;
+        private StreamReader _processError;
+        private CancellationTokenSource _cancellationSource = new CancellationTokenSource();
+        private bool _isClosed = false;
+
+        public CommandShell(string command, string arguments)
+        {
+            ProcessStartInfo processStartInfo = new ProcessStartInfo(command, arguments);
+            processStartInfo.RedirectStandardError = true;
+            processStartInfo.RedirectStandardInput = true;
+            processStartInfo.RedirectStandardOutput = true;
+            processStartInfo.UseShellExecute = false;
+            processStartInfo.CreateNoWindow = true;
+
+            _localProcess = System.Diagnostics.Process.Start(processStartInfo);
+            _localProcess.Exited += OnProcessExited;
+            _localProcess.EnableRaisingEvents = true;
+
+            _stdoutWriter = new StreamWriter(_localProcess.StandardInput.BaseStream, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false), 4096, leaveOpen: true);
+            _processReader = _localProcess.StandardOutput;
+            _processError = _localProcess.StandardError;
+
+            Thread outputThread = new Thread(() => ReadLoop(_processReader, _cancellationSource.Token, (msg) => { OutputReceived?.Invoke(this, msg); }));
+            Thread errorThread = new Thread(() => ReadLoop(_processError, _cancellationSource.Token, (msg) => { ErrorOccured?.Invoke(this, null); }));
+
+            outputThread.Start();
+            errorThread.Start();
+        }
+
+        public event EventHandler<string> OutputReceived;
+        public event EventHandler<int> Closed;
+        public event EventHandler ErrorOccured;
+
+        private static int BUFMAX = 4096;
+        private void ReadLoop(StreamReader reader, CancellationToken token, Action<string> action)
+        {
+            try
+            {
+                while (true)
```

Comment:
>while (true) [](start = 16, length = 12)

What causes this loop to terminate?  The only "return" is commented out.  Does the `ReadAsync` call throw after the underlying command exits?  If so, won't every command execution end with an error event being raised?

Reply:
In the dispose method i call Stream.Close(). I'll change the loop to be `while(!token.IsCancellationRequested)

---
In reply to: [254079310](https://github.com/Microsoft/MIEngine/pull/831#discussion_r254079310) [](ancestors = 254079310)